feature/APPC-1380-AIDL-communication-to-target-billing-available-service

brunovarelaAptoide opened the pull request:
**What does this PR do?**
Adapts the SDK to choose the service to the available billing service that can be: Aptoide,AppcointWallet , Cafe Baazar.
Both the billing and PO projects where changed

**Database changed?**
No

**Where should the reviewer start?**

RepositoryServiceConnection -> android-appcoins-biling
WalletUtils.java -> android-appcoins-biling
WalletUtils.java -> appcoins-ads
PoAManager -> android-appcoiins-billing

**How should this be manually tested?**

Change in build grade the Aptoide_packagename property to .dev.
Use a version from Aptoide with Wallet merged.
Try a normal buy and POA Flux.

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-1380](https://aptoide.atlassian.net/browse/APPC-1380)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new libs, new keys, etc.) 

No


**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

brunovarelaAptoide committed:
Changes

- Created method to check if the wallet billing service available.
- Added String properties

brunovarelaAptoide committed:
Changes

- Binding to aptoideWallet service

brunovarelaAptoide committed:
changes

- Change in how we check the wallet is installed.

brunovarelaAptoide committed:
changes

-changed the way walletIsInstalled is check.
-aadd a verification of package name is null

brunovarelaAptoide committed:
Changes

- Split in 2 functions for 2 diferent behaviours with walletInstalled and not installed.

brunovarelaAptoide committed:
Changes

- Bugfix  when starting appcoins wallet

brunovarelaAptoide committed:
Changes

- Changed POA Wallet install verificatioon
- Changed places where the wallet packagename is used to a generic variable

brunovarelaAptoide committed:
Update gradle.properties

-Changed package name to aptoide prod package name.

brunovarelaAptoide committed:
Changes

Heuristics added to choose between Aptoide+Wallet , Wallet,Bazar Cafe

brunovarelaAptoide committed:
Changes

Added bazar cafe package name

brunovarelaAptoide committed:
Changes

- Added cafe bazaar package name

brunovarelaAptoide committed:
Merge branch 'dev' into feature/APPC-1380-AIDL-communication-to-target-billing-available-service

brunovarelaAptoide committed:
Changes

- Add package name to connection after install Wallet

JDSM01 commented on android-appcoins-billing/build.gradle:
```diff
@@ -26,6 +26,13 @@ android {
       buildConfigField 'String', 'HOST_WS',
           "\"" + project.ROPSTEN_NETWORK_BACKEND_BASE_HOST_WS + "\""
+      buildConfigField 'String', 'SERVICE_BIND_LIST',
+          "\"" + project.APTOIDE_PACKAGE_NAME.replace('\"', '') +
```
This is not giving the aptoide dev package when in debug

JDSM01 replied:
Fixed

JDSM01 committed:
Added aptoide dev package name

JDSM01 commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletUtils.java:
```diff
@@ -1,36 +1,59 @@
+        return true;
+      }
+    }
+    return false;
```
Can't this be simplified with return billingPackageName != null in the end? Same for the other version of the WalletUtils

JDSM01 commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/RepositoryServiceConnection.java:
```diff
@@ -46,19 +45,30 @@ public RepositoryServiceConnection(Context context, ConnectionLifeCycle connecti
     } else {
-      hasWalletInstalled = false;
-      onServiceConnected(new ComponentName("", ""), new IBinderWalletNotInstalled());
+      walletIsNotInstalled();
```
Rename to onWalletNotInstalled()

JDSM01 commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/RepositoryServiceConnection.java:
```diff
@@ -46,19 +45,30 @@ public RepositoryServiceConnection(Context context, ConnectionLifeCycle connecti
     this.listener = listener;
+    if (WalletUtils.hasWalletInstalled()) {
+      String packageName = WalletUtils.getBillingServicePackageName();
+      walletIsInstalled(packageName);
```
Rename to onWalletInstalled

JDSM01 commented on android-appcoins-billing/src/main/AndroidManifest.xml:
```diff
@@ -1,6 +1,7 @@
     package="com.appcoins.billing.sdk">
 
+  <uses-permission android:name="com.appcoins.BILLING"/>
   <uses-permission android:name="com.appcoins.BILLING"/>
```
Why does it appears twice?

JDSM01 commented on appcoins-ads/src/main/java/com/asf/appcoins/sdk/ads/poa/manager/WalletUtils.java:
```diff
@@ -53,14 +57,39 @@ public static boolean hasAptoideInstalled() {
-      return true;
-    } catch (PackageManager.NameNotFoundException e) {
-      return false;
+  public static String chooseServiceToBind(String[] packageNameServices) {
```
can be private

brunovarelaAptoide committed:
Code Review Changes

- Refactor Functions name
- Removed duplicate permission
- Changed a return value

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/RepositoryServiceConnection.java:
```diff
@@ -46,19 +45,30 @@ public RepositoryServiceConnection(Context context, ConnectionLifeCycle connecti
     List<ResolveInfo> intentServices = context.getPackageManager()
         .queryIntentServices(serviceIntent, 0);
     if (intentServices != null && !intentServices.isEmpty()) {
       hasWalletInstalled = true;
```
Does the `hasWalletInstalled` makes sense to be set on the behaviour methods, instead of being set when the if clause in fact asks if the wallet is installer (`WalletUtils.hasWalletInstalled`).

Also if we know that the wallet is installed is it possible for the intent service to be null or empty in the `walletInstalledBehaviour`? If so and since that is the main condition for the proceed with the behaviour the we should use the  `walletInstalledBehaviour` logic only instead of having one if clause in the `startConnection` and then a second in the `walletNotInstalledBehaviour`?

brunovarelaAptoide committed:
Code review changes

- Changed the organization of the connection structure in RepositoryServiceConnection.java

JoaoRaimundo commented on android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/RepositoryServiceConnection.java:
```diff
@@ -46,22 +43,27 @@ public RepositoryServiceConnection(Context context, ConnectionLifeCycle connecti
   }
 
+  private void walletNotInstalledBehaviour() {
+    hasWalletInstalled = false;
```
This attribution is redundant

brunovarelaAptoide committed:
Code review changes

- Removed reduntant variable.

brunovarelaAptoide committed:
code review changes

- removed bug from the algorithm to choose the service to bind.
- changed logic to from a array to a arraylist
